Add peertablesortproxy module

hebasto opened the pull request:
The "Peers" table in the "Node" window does not hold multiple selection after sorting.

This PR introduces a `QSortFilterProxyModel` subclass, that is a standard Qt [practice](https://doc.qt.io/qt-5/model-view-programming.html#custom-sorting-models) for such cases.

Now the sorting code is encapsulated into the dedicated Qt class, and we do not need to maintain it.

Fixes #283 (additionally).

---

On **master** (7ae86b3c6845873ca96650fc69beb4ae5285c801):
- rows are sorted by "Ping", and a selection is made
![Screenshot from 2020-11-28 22-53-11](https://user-images.githubusercontent.com/32963518/100525900-96eaed00-31cc-11eb-86e7-72ede3b8b33c.png)

- rows are sorted by "NodeId", and the previous selection is _lost_
![Screenshot from 2020-11-28 22-53-21](https://user-images.githubusercontent.com/32963518/100525904-9c483780-31cc-11eb-957c-06f53d7d31ab.png)

With **this PR**:
- rows are sorted by "Ping", and a selection is made
![Screenshot from 2020-11-28 22-39-41](https://user-images.githubusercontent.com/32963518/100525776-06aca800-31cc-11eb-8c4e-9c6566fe80fe.png)

- rows are sorted by "NodeId", and the row are still selected
![Screenshot from 2020-11-28 22-39-53](https://user-images.githubusercontent.com/32963518/100525791-2348e000-31cc-11eb-8b78-716a5551d7ec.png)

MarcoFalke commented:
Is this a refactor or feature or bugfix?

hebasto commented:
> Is this a refactor or feature or bugfix?

A bugfix, I think.

The correct handling of multiple selections during sorting by a column could be observer in the transaction table.

jonasschnelli commented:
No sure if this is worth the effort.
No strong opinion though.
Please rebase If this should be dragged along.

hebasto commented:
> Please rebase If this should be dragged along.

Rebased 0a03c82cefab82e273c5e76a3a3b93ae6a75df54 -> b119ae0c82c6f36afdf3eefd89b687ab78a9e34a ([pr18.01](https://github.com/hebasto/gui/commits/pr18.01) -> [pr18.02](https://github.com/hebasto/gui/commits/pr18.02)).

jarolrod reviewed (commented):
Tested ACK b119ae0 on Arch Linux and macOS 10.15.7. Code looks good and PR does what it says it will do. 

The behavior this PR introduces of showing nothing is a better behavior than the default functionality. I think an even better behavior would be to, as you select a group of peers, show the last peer that you selected in that selection group.

hebasto commented:
@jarolrod 
> The behavior this PR introduces of showing nothing is a better behavior than the default functionality.

Do you mean #13?

jarolrod commented:
> @jarolrod
> 
> > The behavior this PR introduces of showing nothing is a better behavior than the default functionality.
> 
> Do you mean #13?

Oh yes, sorry. Overlooked that this builds upon that PR.

hebasto commented:
Rebased b119ae0c82c6f36afdf3eefd89b687ab78a9e34a -> ab070cf39a6756723a4efc17e6e98de9d47425d5 ([pr18.02](https://github.com/hebasto/gui/commits/pr18.02) -> [pr18.03](https://github.com/hebasto/gui/commits/pr18.03)) due to the conflict #13.

hebasto commented:
The OP updated with screenshots.

hebasto commented:
Rebased ab070cf39a6756723a4efc17e6e98de9d47425d5 -> 5745711a8b2641fe727b98fa93425f4b0f4e4beb ([pr18.03](https://github.com/hebasto/gui/commits/pr18.03) -> [pr18.04](https://github.com/hebasto/gui/commits/pr18.04)) on top of the recent changes in p2p logic and CI.

hebasto commented:
Converting to draft as a bug was discovered while testing in the light of #158.

hebasto commented:
Updated 6f1ccc74b260413ad6750a693e8e0bdc36881300 -> 6c273034da293f6d4ac691959c5c6435f67200be ([pr18.05](https://github.com/hebasto/gui/commits/pr18.05) -> [pr18.06](https://github.com/hebasto/gui/commits/pr18.06)):
- fixed a bug with the detailed widget update
- rebased on top of #161

Please review #161 first :)

hebasto commented:
@jonasschnelli 
> No sure if this is worth the effort.

Benefits:
- less lines of code (+109 / -175)
- using a proper abstraction layer for sorting makes the code more readable and maintainable

hebasto commented:
Rebased 6c273034da293f6d4ac691959c5c6435f67200be -> 9bc2004db26a3a956256f8fd9b05e7a3e9e21aca ([pr18.06](https://github.com/hebasto/gui/commits/pr18.06) -> [pr18.07](https://github.com/hebasto/gui/commits/pr18.07)) due to the conflict with #162.

hebasto commented:
Rebased 9bc2004db26a3a956256f8fd9b05e7a3e9e21aca -> fb6901787d7eafec5bc17d5f0c271935ac1e1c17 ([pr18.07](https://github.com/hebasto/gui/commits/pr18.07) -> [pr18.08](https://github.com/hebasto/gui/commits/pr18.08)) due to the conflict with #161.

jarolrod reviewed (commented):
re-ack fb6901787d7eafec5bc17d5f0c271935ac1e1c17, tested on macOS 11.1 with Qt 5.15.2

@jonasschnelli in regards to [comment](https://github.com/bitcoin-core/gui/pull/18#issuecomment-715217620): If I wanted to see how certain peers rank based on different metrics, it would be nice to have the peers I have in interest remain selected as I change the sorting. So, this PR does add value.

Below are screenshots showing the behavior between `master` and this `PR`

### Master
**Master:** Select three peers with default sorting (by Peer ID)
<img width="896" alt="Screen Shot 2021-01-18 at 8 15 34 PM" src="https://user-images.githubusercontent.com/23396902/104975974-1ecdc600-59ca-11eb-9070-c9fc4dd0daba.png">

**Master:** Change to sorting by Ping -> peers of interest are no longer selected
<img width="897" alt="Screen Shot 2021-01-18 at 8 15 43 PM" src="https://user-images.githubusercontent.com/23396902/104976045-44f36600-59ca-11eb-8b43-2919946c7d3d.png">

### PR
**PR:**  Select three peers with default sorting (by Peer ID)
<img width="889" alt="Screen Shot 2021-01-18 at 7 54 22 PM" src="https://user-images.githubusercontent.com/23396902/104976115-6eac8d00-59ca-11eb-8ca7-5bee7d1edb71.png">

**PR:** Change to sorting by Ping -> Peers remain selected
<img width="893" alt="Screen Shot 2021-01-18 at 7 54 30 PM" src="https://user-images.githubusercontent.com/23396902/104976148-85eb7a80-59ca-11eb-8132-5547fdba7c8a.png">

hebasto commented:
Rebased fb6901787d7eafec5bc17d5f0c271935ac1e1c17 -> 4a032f64e45c7d78a378057e878acc0b101467a6 ([pr18.08](https://github.com/hebasto/gui/commits/pr18.08) -> [pr18.09](https://github.com/hebasto/gui/commits/pr18.09)) due to the conflict with #180.

jarolrod reviewed (commented):
re-ack 4a032f64e45c7d78a378057e878acc0b101467a6

hebasto commented:
Rebased 4a032f64e45c7d78a378057e878acc0b101467a6 -> 8d4a682791bedcc9cf471e9d77e77ec46d0bc314 ([pr18.09](https://github.com/hebasto/gui/commits/pr18.09) -> [pr18.10](https://github.com/hebasto/gui/commits/pr18.10)) due to the conflict with #179.

jarolrod reviewed (commented):
re-ACK 8d4a682791bedcc9cf471e9d77e77ec46d0bc314, tested on macOS 11.1 Qt 5.15.2 after rebase

hebasto commented:
Rebased 8d4a682791bedcc9cf471e9d77e77ec46d0bc314 -> 56e5f5aefce6df0e54fbbd556d7bcdca3f167d12 ([pr18.10](https://github.com/hebasto/gui/commits/pr18.10) -> [pr18.11](https://github.com/hebasto/gui/commits/pr18.11)) due to the conflict with https://github.com/bitcoin/bitcoin/pull/21015.

hebasto commented:
Rebased 56e5f5aefce6df0e54fbbd556d7bcdca3f167d12 -> 5a4a15d2b4456272fd8aa080195f40a09576ae01 ([pr18.11](https://github.com/hebasto/gui/commits/pr18.11) -> [pr18.12](https://github.com/hebasto/gui/commits/pr18.12)) due to the conflict with #166.

jarolrod commented:
re-ACK 5a4a15d2b4456272fd8aa080195f40a09576ae01, tested on macOS 11.2 Qt 5.15.2 after rebase

jarolrod reviewed (commented):
re-ACK 5a4a15d2b4456272fd8aa080195f40a09576ae01, tested on macOS 11.2 Qt 5.15.2 after rebase

rebroad commented:
seems like a fair bit of code for functionality which I'd expect by default.... nice work though.

promag commented on src/qt/clientmodel.cpp:
```diff
@@ -184,6 +189,11 @@ PeerTableModel *ClientModel::getPeerTableModel()
     return peerTableModel;
 }
 
+PeerTableSortProxy* ClientModel::peerTableSortProxy()
```
778a64af209e4fa692a3aca8376ba1bd5e1af881

I don't think this belongs here, should be a member of `RPCConsole` - usually, sort/filter models are instanced for each view unless you want to share the same sorting/filter result in multiple views (I know only it's only used in `RPCConsole`).

hebasto replied:
From the `##bitcoin-core-gui` IRC channel:
> \<promag> https://github.com/bitcoin-core/gui/pull/18 seems ready
> \<promag> https://github.com/bitcoin-core/gui/pull/18#discussion_r616065705 can be addressed later

promag reviewed (commented):
Concept ACK, I think this is the best approach to keep a model sorted and eventually filtered.

Should `setDynamicSortFilter(true)`?

promag reviewed (commented):
Tested ACK 5a4a15d2b4456272fd8aa080195f40a09576ae01.

#283 is fixed, selected cell os copied with copy shortcut.

hebasto commented:
@laanwj 
> Concept ACK, haven't reviewed the code in detail but conceptually this is the proper Qt way of doing this.

Do you mind to have another look at this PR to promote your "Concept ACK"?
